src/v/cluster/shard_table.h

```diff
@@ -49,7 +50,8 @@ class shard_table final {
     /**
      * \brief Lookup the owning shard for an ntp.
      */
-    std::optional<ss::shard_id> shard_for(const model::ntp& ntp) {
+    template<model::any_ntp T>
+    std::optional<ss::shard_id> shard_for(const T& ntp) {
```

Comment:
Typical of this transformation: take any_ntp (concept for ntp + ktp + ktp_with_hash) instead of just ntp and then heterogeneous lookup does the rest.